### PR TITLE
Add ability to have multiple loss functions

### PR DESCRIPTION
**Pull request by kaminow**
We want to be able to incorporate multiple loss functions into our training, eg standard affinity target loss + loss penalty for predicting values outside a normal pIC50 range. This PR makes that infrastructure change, and also improves how we're tracking loss values and model predictions over training. Also introduce some new loss functions.

- [x] Allow multiple loss functions in `Trainer`/CLI
- [x] Add range penalty loss function
- [x] Add loss function for correctly ranking poses
- [x] Add pose ranking to the dataset loading (labels for above loss function in training)
- [x] Add `asapdiscovery.ml.schema.TrainingPrediction` and `asapdiscovery.ml.schema.TrainingPredictionTracker` to organize how we track predictions
- [x] New `ml.viz` module with common visualization stuff (loss curves, prediction plots, etc)

### CODE REVIEW

**Comment by kaminow** [comment]
Note that this shouldn't be merged until choderalab/mtenn#66 is merged and a new `mtenn` release cut and pinned here

**Comment by kaminow** [comment]
@hmacdope I think this is ready for a review. happy to meet at some point to go over it synchronously as well if you want

File: asapdiscovery-ml/asapdiscovery/ml/tests/test_ml_cli.py [hunk 1]
```diff
@@ -464,24 +465,24 @@ def test_build_and_train_graph(exp_file, tmp_path):
...
     # Load and check stuff
-    loss_dict = json.loads(loss_dict_file.read_text())
-    assert {"train", "test", "val"} == set(loss_dict.keys())
-    print({k: len(v) for k, v in loss_dict.items()}, flush=True)
-    assert len(loss_dict["train"]) == 8
-    assert len(loss_dict["val"]) == 1
-    assert len(loss_dict["test"]) == 1
+    pred_tracker = TrainingPredictionTracker(**json.loads(tpt_path.read_text()))
+    assert {"train", "test", "val"} == set(pred_tracker.split_dict.keys())
+    print({k: len(v) for k, v in pred_tracker.split_dict.items()}, flush=True)
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/tests/test_ml_cli.py:482]: Remove print

File: asapdiscovery-ml/asapdiscovery/ml/viz.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+        ):
+            stats_str = (
+                f"{stat_label}: "
+                f"{stats_dict[lab]['test'][stat]['value']:0.2f}"
+                f"$_{{{stats_dict[lab]['test'][stat]['95ci_low']:0.2f}}}"
+                f"^{{{stats_dict[lab]['test'][stat]['95ci_high']:0.2f}}}$"
+            )
+            stats_text.append(stats_str)
+
+        # ax.text(
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/viz.py:243]: Clean up

**Review by hmacdope — CHANGES_REQUESTED** [review]
Few small things, although fundamentally I am relying on you having tested this thoroughly and maintaining it, as I don't understand every part.

Let me also check the skynet code works in its entirety as well.

**Comment by hmacdope** [comment]
It also looks like you will need to fix some mlops stuff here also before we merge.

File: asapdiscovery-ml/asapdiscovery/ml/cli_args.py [hunk 3]
```diff
@@ -1292,9 +1295,12 @@ def batch_size(func):
     )(func)
 
 
-def target_prop(func):
+def target_props(func):
     return click.option(
-        "--target-prop", type=str, help="Target property to train against."
+        "--target-prop",
+        type=str,
+        multiple=True,
```
[the remark this answers is not in the document]
Comment (kaminow) [reply asapdiscovery-ml/asapdiscovery/ml/cli_args.py:1302]: I guess this is really only like half a functionality at this point. the model will still only give one prediction for a given input, no matter how many target prop values you have. this can let you train against two values simultaneously, but it would only make sense at all if they were related (eg pIC50 and pKi), but now that I'm thinking about it idk how much sense it makes to have this functionality at all without having the model be able to make multiple predictions. do you think it's ok to leave in for now since it still works with just one prop specified, or should we get rid of it for the time being to avoid confusing people?

File: asapdiscovery-ml/asapdiscovery/ml/dataset.py [hunk 4]
```diff
@@ -51,6 +52,9 @@ def from_complexes(cls, complexes: list[Complex], exp_dict={}, ignore_h=True):
         DockedDataset
         """
 
+        if exp_dict is None:
```
[the remark this answers is not in the document]
Comment (kaminow) [reply asapdiscovery-ml/asapdiscovery/ml/dataset.py:55]: it's a fun little bit of Python [nonsense](https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python)

File: asapdiscovery-ml/asapdiscovery/ml/cli_sweep.py [hunk 5]
```diff
@@ -693,12 +722,10 @@ def sweep_e3nn(
         "enforce_one": enforce_one,
         "rand_seed": ds_rand_seed,
     }
-    loss_config = {
-        "cache": loss_config_cache,
-        "overwrite_cache": overwrite_loss_config_cache,
-        "loss_type": loss_type,
-        "semiquant_fill": semiquant_fill,
-    }
+    loss_configs = [
```
[the remark this answers is not in the document]
Comment (kaminow) [reply asapdiscovery-ml/asapdiscovery/ml/cli_sweep.py]: put it in `cli_args.py`, not sure if that's the right place but it's somewhere that won't result in a circular import anywhere.  if you think that's bad though I can make a `utils` file and move it there

File: asapdiscovery-ml/asapdiscovery/ml/config.py [hunk 6]
```diff
@@ -522,15 +522,25 @@ def build(self):
...
+                    pose["z"] = pose["lig"].reshape((-1, 1)).float()
+            else:
+                # Check if this data has already been adjusted
+                if data["z"].is_floating_point():
+                    # Assume it'll only be floats if we've already run this function
+                    continue
 
-            pose["x"] = torch.nn.functional.one_hot(pose["z"] - 1, 100).float()
-            pose["z"] = pose["lig"].reshape((-1, 1)).float()
+                data["x"] = torch.nn.functional.one_hot(data["z"] - 1, 100).float()
```
[the remark this answers is not in the document]
Comment (kaminow) [reply asapdiscovery-ml/asapdiscovery/ml/config.py:544]: yeah good point, fixed so it doesn't

File: asapdiscovery-data/asapdiscovery/data/scripts/cdd_to_schema.py [hunk 7]
```diff
@@ -12,38 +12,82 @@
...
-    parser.add_argument("-json", required=True, help="Output JSON file.")
-    parser.add_argument("-csv", help="Output CSV file.")
-    parser.add_argument(
-        "-type",
-        default="std",
-        help=(
-            "What type of data is " "being loaded (std: standard, ep: enantiomer pairs)"
-        ),
-    )
+@click.command()
```
[the remark this answers is not in the document]
Comment (kaminow) [reply asapdiscovery-data/asapdiscovery/data/scripts/cdd_to_schema.py]: done and done

File: asapdiscovery-ml/asapdiscovery/ml/schema.py [hunk 8]
```diff
@@ -0,0 +1,934 @@
...
+                confidence_level=0.95,
+                paired=True,
+            ).confidence_interval
+            stats_dict[sp]["tau"] = {
+                "value": tau,
+                "95ci_low": conf_interval.low,
+                "95ci_high": conf_interval.high,
+            }
+
+        return stats_dict
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/schema.py:934]: This is a lot of code, is there any chance we can drop some? I have no ability to really vouch for / test it so perhaps cutting down volume if possible.

File: asapdiscovery-ml/asapdiscovery/ml/tests/test_ml_cli.py [hunk 9]
```diff
@@ -392,8 +393,8 @@ def test_build_trainer_visnet(exp_file, docked_files, tmp_path):
             tmp_path / "ds_cache.pkl",
             "--ds-config-cache",
             tmp_path / "ds_config_cache.json",
-            "--loss-type",
-            "mse_step",
+            "--loss",
+            "loss_type:mse_step",
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/tests/test_ml_cli.py:397]: Is the default still preserved wrt old CLI?

File: asapdiscovery-ml/asapdiscovery/ml/tests/test_training_prediction_schema.py [hunk 10]
```diff
@@ -0,0 +1,607 @@
...
+        loss_val=30.0,
+        **identifiers[3],
+        loss_config=loss_configs[0],
+    )
+
+    loss_dict = tp_tracker.get_losses(agg_compounds=True, agg_losses=True)
+
+    assert set(loss_dict.keys()) == {"train"}
+
+    # Weighted mean across loss configs and regular mean across compounds
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/tests/test_training_prediction_schema.py:601]: Ok you wrote very thorough tests.

File: asapdiscovery-ml/asapdiscovery/ml/trainer.py [hunk 11]
```diff
@@ -775,25 +945,75 @@ def train(self):
...
+                        target_val=target,
+                        in_range=in_range,
+                        uncertainty=uncertainty,
+                        loss_config=loss_config,
+                        loss_weight=loss_wt,
+                    )
+
+                # If all target props were missing, there's no backprop to do
+                if not loss.requires_grad:
+                    continue
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/trainer.py:1012]: Gonna trust you on this.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Few changes but almost there.

File: asapdiscovery-ml/asapdiscovery/ml/cli.py [hunk 12]
```diff
@@ -257,12 +255,10 @@ def build_gat(
         "enforce_one": enforce_one,
         "rand_seed": ds_rand_seed,
     }
-    loss_config = {
-        "cache": loss_config_cache,
-        "overwrite_cache": overwrite_loss_config_cache,
-        "loss_type": loss_type,
-        "semiquant_fill": semiquant_fill,
-    }
+    loss_configs = [
```
[the remark this answers is not in the document]
Comment (kaminow) [reply asapdiscovery-ml/asapdiscovery/ml/cli.py:258]: oh weird I swear I replaced all of these lol, I'll make sure to redo it

File: asapdiscovery-ml/asapdiscovery/ml/tests/test_ml_cli.py (hunk 9 above)
Comment (kaminow) [reply asapdiscovery-ml/asapdiscovery/ml/tests/test_ml_cli.py:397]: I don't think there were any defaults before? you've always had to specify which loss function you want to use

**Review by hmacdope — APPROVED** [review]
LGTM!
